### PR TITLE
OD-16387 New columns for all list views

### PR DESCRIPTION
**Pull request by dmitars**
…ith bad performance

### CODE REVIEW

File: server/src/main/groovy/ish/oncourse/server/cayenne/CourseClass.groovy [hunk 1]
```diff
@@ -738,6 +738,14 @@ class CourseClass extends _CourseClass implements CourseClassTrait, Queueable, N
 		return super.getIsActive()
 	}
 
+	/**
+	 * @return true if class will begin in future
+	 */
+	@API
+	Boolean getIsFuture() {
+		return startDateTime == null || startDateTime.after(new Date());
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/CourseClass.groovy:745]: let's check startDateTime != null && startDateTime.after(new Date());
If class has no startDateTime - it is unscheduled or distant learning

File: server/src/main/groovy/ish/oncourse/server/cayenne/MembershipProduct.groovy [hunk 2]
```diff
@@ -63,6 +65,22 @@ class MembershipProduct extends _MembershipProduct {
 		return super.getDiscountMemberships()
 	}
 
+	/**
+	 * @return number of active product items
+	 */
+	@API
+	Long getActiveMembershipsCount() {
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/MembershipProduct.groovy:71]: What the purpose to keep interface in cayenne but implementation in trait, just move that methods it trait complitely

File: server/src/main/groovy/ish/oncourse/server/cayenne/MembershipProduct.groovy [hunk 3]
```diff
@@ -63,6 +65,22 @@ class MembershipProduct extends _MembershipProduct {
...
+		return getTotalActiveMembershipsCount()
+	}
+
+	/**
+	 * @return number of sold product items
+	 */
+	@API
+	Long getSoldCount() {
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/MembershipProduct.groovy:78]: same problem

File: server/src/main/groovy/ish/oncourse/server/cayenne/Outcome.groovy [hunk 4]
```diff
@@ -389,6 +390,22 @@ class Outcome extends _Outcome implements Queueable, OutcomeTrait {
 		return LocalDateUtils.dateToValue(calculateEndDate(Boolean.FALSE))
 	}
 
+	/**
+	 * @return percent of attended hours; if enrolment not set, returns 0 by default
+	 */
+	@API
+	Long getAttendedPercent() {
+		return enrolment != null ? getPresentAttendancePercent() : 0
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/Outcome.groovy:397]: Move in trait please

File: server/src/main/groovy/ish/oncourse/server/cayenne/OutcomeTrait.groovy [hunk 5]
```diff
@@ -45,4 +45,133 @@ trait OutcomeTrait {
             return []
         }
     }
+
+    Long getPresentAttendancePercent() {
+        return Math.round(getAttendedAttendancesDuration().doubleValue() / getMaxPossibleAttendedDuration().doubleValue() * 100)
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/OutcomeTrait.groovy:50]: What if the divider is 0 here?

File: server/src/main/groovy/ish/oncourse/server/cayenne/OutcomeTrait.groovy [hunk 6]
```diff
@@ -45,4 +45,133 @@ trait OutcomeTrait {
...
+
+    Long getMarkedAssessmentPercent() {
+        return Math.round((double) getMarkedAssessmentsCount() / getOutcomeAssessments().size())
+    }
+
+    BigDecimal getMaxPossibleAttendedDuration() {
+        return absentAttendancesDuration + attendedAttendancesDuration + notMarkedAttendancesDuration + futureTimetable
+    }
+
+    BigDecimal getNotMarkedAttendancesDuration() {
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/OutcomeTrait.groovy:61]: I'd like you to write the unit test for all OutcomeTrait methods please

File: server/src/main/groovy/ish/oncourse/server/cayenne/RoomTrait.groovy [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package ish.oncourse.server.cayenne
+
+trait RoomTrait {
+
+    abstract List<CourseClass> getCourseClasses()
+
+    Long getActiveClassesCount(){
+        def classes = getCourseClasses()
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/RoomTrait.groovy:16]: That is dangerous method invocation (potentially a lots of classes can be here), replace with select count please

File: server/src/main/groovy/ish/oncourse/server/cayenne/RoomTrait.groovy [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        return classes == null? 0: classes.findAll {courseClass -> courseClass.isActive}.size()
+    }
+
+    Long getFutureClassesCount(){
+        def classes = getCourseClasses()
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/cayenne/RoomTrait.groovy:21]: Same problemm

**Review by ArtyomKravchenko — CHANGES_REQUESTED** [review]

**Review by ArtyomKravchenko — APPROVED** [review]

File: server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy [hunk 9]
```diff
@@ -127,6 +129,8 @@ class DefaultUserPreference {
                 new ColumnDTO(title: 'Code', attribute: Module.NATIONAL_CODE.name, sortable: true, width: W100, visible: true),
                 new ColumnDTO(title: 'Title', attribute: Module.TITLE.name, sortable: true, width: W100, visible: true),
                 new ColumnDTO(title: 'Is offered', attribute: Module.IS_OFFERED.name, sortable: true, width: W100, visible: true, type: ColumnTypeDTO.BOOLEAN),
+                new ColumnDTO(title: 'Credit points', attribute: Module.CREDIT_POINTS.name, sortable: true, width: W100, visible: false, type: ColumnTypeDTO.MONEY),
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy:133]: Money is wrong here

File: server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy [hunk 10]
```diff
@@ -147,7 +151,9 @@ class DefaultUserPreference {
                 new ColumnDTO(title: 'Local timezone', attribute: Site.LOCAL_TIMEZONE.name, sortable: true, width: W200, visible: false),
                 new ColumnDTO(title: 'Administration centre', attribute: Site.IS_ADMINISTRATION_CENTRE.name, sortable: true, width: W200, visible: true, type: ColumnTypeDTO.BOOLEAN),
                 new ColumnDTO(title: 'Shown on web', attribute: Site.IS_SHOWN_ON_WEB.name, sortable: true, width: W200, visible: false, type: ColumnTypeDTO.BOOLEAN),
-                new ColumnDTO(title: 'Is virtual', attribute: Site.IS_VIRTUAL.name, sortable: true, width: W200, visible: false, type: ColumnTypeDTO.BOOLEAN)
+                new ColumnDTO(title: 'Is virtual', attribute: Site.IS_VIRTUAL.name, sortable: true, width: W200, visible: false, type: ColumnTypeDTO.BOOLEAN),
+                new ColumnDTO(title: 'Active classes', attribute: Site.ACTIVE_CLASSES_COUNT_KEY, sortable: false, width: W200, visible: false, prefetches: [Site.ROOMS.path().toString()]),
+                new ColumnDTO(title: 'Future classes', attribute: Site.FUTURE_CLASSES_COUNT_KEY, sortable: false, width: W200, visible: false),
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy:158]: Room prefetch here

File: server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy [hunk 11]
```diff
@@ -264,8 +274,10 @@ class DefaultUserPreference {
...
                 new ColumnDTO(title: 'Created', attribute: Payslip.CREATED_ON.name, sortable: true, width: W200, visible: true, type: ColumnTypeDTO.DATE),
                 new ColumnDTO(title: 'Status', attribute: Payslip.STATUS.name, sortable: true, width: W200, visible: true),
-                new ColumnDTO(title: 'Type', attribute: Payslip.PAY_TYPE.name, sortable: true, width: W200, visible: true)
-
+                new ColumnDTO(title: 'Type', attribute: Payslip.PAY_TYPE.name, sortable: true, width: W200, visible: true),
+                new ColumnDTO(title: 'Pay type', attribute: Payslip.CONTACT.dot(Contact.TUTOR).dot(Tutor.PAY_TYPE).name, sortable: false, width: W200, visible: false, prefetches: [Payslip.CONTACT.dot(Contact.TUTOR).path().toString()]),
+                new ColumnDTO(title: 'Budget total', attribute: Payslip.BUDGET_TOTAL_KEY, sortable: false, width: W100, visible: false, prefetches: [Payslip.PAYLINES.path().toString()]),
+                new ColumnDTO(title: 'Paid total', attribute: Payslip.PAID_TOTAL_KEY, sortable: false, width: W100, visible: false),
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy:285]: Prefetch payslip here

File: server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy [hunk 12]
```diff
@@ -462,7 +475,22 @@ class DefaultUserPreference {
...
                         visible: false,
                         system: true),
-
+                new ColumnDTO(title: 'Enrolment type',
+                        attribute: Course.ENROLMENT_TYPE.name,
+                        sortable: true,
+                        width: W100,
+                        visible: false,
+                        system: true),
+                new ColumnDTO(title: 'Data collection',
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy:494]: Add fieldConfigurationScheme prefetch

File: server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy [hunk 13]
```diff
@@ -656,6 +693,7 @@ class DefaultUserPreference {
                 new ColumnDTO(title: 'Code', attribute: Assessment.CODE.name, sortable: true, width: W200, visible: true),
                 new ColumnDTO(title: 'Name', attribute: Assessment.NAME.name, sortable: true, width: W200, visible: true),
                 new ColumnDTO(title: 'Active', attribute: Assessment.ACTIVE.name, sortable: true, width: W200, visible: false, system: true, type: ColumnTypeDTO.BOOLEAN),
+                new ColumnDTO(title: 'Grading type', attribute: Assessment.GRADING_TYPE.dot(GradingType.TYPE_NAME).name, sortable: false, width: W200, visible: false),
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy:707]: GradingType prefetch

File: server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy [hunk 14]
```diff
@@ -689,6 +727,8 @@ class DefaultUserPreference {
                 ),
                 new ColumnDTO(title: 'Submitted on', attribute: AssessmentSubmission.SUBMITTED_ON.name, sortable: true, width: W200, visible: true, type: ColumnTypeDTO.DATETIME),
                 new ColumnDTO(title: 'Marked on', attribute: AssessmentSubmission.MARKED_ON.name, sortable: true, width: W200, visible: true, type: ColumnTypeDTO.DATETIME),
+                new ColumnDTO(title: 'Grade', attribute: AssessmentSubmission.GRADE.name, sortable: true, width: W200, visible: false),
+                new ColumnDTO(title: 'Assessor', attribute: AssessmentSubmission.MARKED_BY.dot(Contact.FULL_NAME_KEY).name, sortable: false, width: W200, visible: false),
```
Comment (ArtyomKravchenko) [inline server/src/main/groovy/ish/oncourse/server/preference/DefaultUserPreference.groovy:742]: Prefetch here

**Review by ArtyomKravchenko — CHANGES_REQUESTED** [review]

**Review by ArtyomKravchenko — APPROVED** [review]

**Review by ArtyomKravchenko — APPROVED** [review]

**Review by ArtyomKravchenko — DISMISSED** [review]

**Review by GFilipovich — APPROVED** [review]

**Pull request merged**